### PR TITLE
xds: reimplement stats management

### PR DESCRIPTION
**Pull request by voidzcy**
This PR include two major changes:

1. Implemented the new stats objects: `ClusterDropStats` and `ClusterLocalityStats`, which match [C-core's implementation](https://github.com/grpc/grpc/blob/master/src/core/ext/xds/xds_client_stats.h). They expose very small interfaces: `Record()` and `Release()`. 

   The two XdsClient APIs for accessing stats objects are
      - `addClusterDropStats(String clusterName, String edsServiceName)` 
      - `addClusterLocalityStats(String clusterName, String edsServiceName, Locality locality)`

   Note there are no `Remove()` methods since stats object consumers can just call `Release()` to release the usages of stats objects (in C-core, this is done in stats object's [destructor](https://github.com/grpc/grpc/blob/e6c5554b0e6bda932c6f0fca29899d6df1f6bcd5/src/core/ext/xds/xds_client_stats.cc#L71), stats object consumers do not interact with XdsClient's `Remove()` directly).

   Note I deleted code for recording backend metrics in per-locality stats objects. The latest xDS design does not include handlings for backend metrics at all. C-core doesn't have implementation for it as well. I believe it will be redesigned in the future if we want that feature. Anyway, it's dead code today and annoying (tests) to be there.

2. The LRS LB policy is eliminated, as a requirement for the c2p affinity design. The endpoint addresses resolved in ClusterResolverLoadBalancer will attach the locality in each address attributes. In ClusterImplLoadBalancer, its helper's `createSubchannel()` will populate the address locality and then call `XdsClient.addClusterLocalityStats(...)` to obtain the per-locality stats object for recording RPCs. This stats object is attached to the created _subchannel's attribute_. Therefore, ClusterImplLoadBalancer receives Picker update from its child LB policy, the Picker's subchannel will always have the per-locality stats object attached. `Helper.pickSubchannel(...)` will populate the per-locality stats object and wrap it into the stream tracer for counting RPCs. Note the subchannel's `shutdown()` is wrapped to call the stats object's `Release()`.

-------------------
Suggested review order:
- `LoadStatsManager2`, which includes implementations for `ClusterDropStats` and `ClusterLocalityStats`.
- `XdsClient`, `ClientXdsClient`, which defines the APIs for accessing stats objects from LB policies. The API calls are actually delegated to `LoadStatsManager2`, plus it turns on load reporting.
- `ClusterResolverLoadBalancer`:
   - Endpoint addresses (EAG) are attached with the locality that the address is in. For addresses resolved by DNS, a locality with empty region/zone/subzone is used.
   - LRS LB policy is removed from the LB policy configs it generates for the downstream LB tree. That is, the child LB policies of weighted_target are "round_robin" (`endpointPickingPolicy` from its config) or "pick_first" (for endpoints resolved by DNS).
- `ClusterImplLoadBalancer`:
   - Obtains the `ClusterDropStats` object from XdsClient and uses it to record dropped requests (same as before for `LoadStatsStore`).
   - Most important changes are in its Helper:
      - `createSubchannel()` wraps the created Subchannel by putting a `ClusterLocalityStats` to the Subchannel's attributes (first populates the locality from the subchannel's addresses, then obtain the stats object from XdsClient). Note the Subchannel's `shutdown()` is intercepted to _release the stats object_. So the reference lifecycle of the `ClusterLocalityStats` object is same as the `Subchannel`.
      - `updateBalancingState()` gets the Picker propagated from the child LB policy. The Picker is wrapped so that a `ClientStreamTracer.Factory` is put into the PickResult. Note the PickResult picked by the Picker propagated from the child LB policy has the `ClusterLocalityStats` object in its attributes. That stats object is populated to be wrapped into the `ClientStreamTracer.Factory`.
- Tests changes and other deleted classes.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/LoadStatsManager2.java [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+  class ClusterDropStats {
+    private final String clusterName;
+    @Nullable
+    private final String edsServiceName;
+    private final AtomicLong uncategorizedDrops = new AtomicLong();
+    private final ConcurrentMap<String, AtomicLong> categorizedDrops = new ConcurrentHashMap<>();
+    private final Stopwatch stopwatch;
+
+    private ClusterDropStats(
+        String clusterName, @Nullable String edsServiceName, Stopwatch stopwatch) {
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LoadStatsManager2.java:253]: Now ClusterDropStats creation is LoadStatsManager2 internal, the outer class has the full control for Stopwatch instances to be passed in.

File: xds/src/main/java/io/grpc/xds/LoadStatsManager2.java [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    return ref.get();
+  }
+
+  private synchronized void releaseClusterDropCounter(
+      String cluster, @Nullable String edsServiceName) {
+    checkState(allDropStats.containsKey(cluster)
+            && allDropStats.get(cluster).containsKey(edsServiceName),
+        "stats for cluster %s, edsServiceName %s not exits", cluster, edsServiceName);
+    ReferenceCounted<ClusterDropStats> ref = allDropStats.get(cluster).get(edsServiceName);
+    ref.release();
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LoadStatsManager2.java:91]: Calling stats object's `release()` does not immediately deletes the counter. It just decrements the ref count. The counter will be discarded after next load report. See line 162.

**Comment by voidzcy** [comment]
Ping.

File: xds/src/main/java/io/grpc/xds/XdsLbPolicies.java [hunk 3]
```diff
@@ -24,7 +24,6 @@
   static final String PRIORITY_POLICY_NAME = "priority_experimental";
   static final String CLUSTER_IMPL_POLICY_NAME = "cluster_impl_experimental";
   static final String WEIGHTED_TARGET_POLICY_NAME = "weighted_target_experimental";
-  static final String LRS_POLICY_NAME = "lrs_experimental";
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsLbPolicies.java:27]: EDS can also be removed.

File: xds/src/main/java/io/grpc/xds/LoadStatsManager2.java [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+     * LoadStatsManager2#getClusterLocalityStats}). The object may still be
+     * recording loads after this method, but there is no guarantee loads recorded after this
+     * point will be included in load reports.
+     */
+    void release() {
+      LoadStatsManager2.this.releaseClusterLocalityLoadCounter(
+          clusterName, edsServiceName, locality);
+    }
+
+    private synchronized ClusterLocalityStatsSnapshot snapshot() {
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LoadStatsManager2.java]: You are right. Fixed here and ClusterDropStats.

File: xds/src/main/java/io/grpc/xds/LoadStatsManager2.java [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+      res.addAll(getClusterStatsReports(cluster));
+    }
+    return Collections.unmodifiableList(res);
+  }
+
+  /**
+   * Recorder for dropped requests. One instance per cluster with edsServiceName.
+   */
+  @ThreadSafe
+  class ClusterDropStats {
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LoadStatsManager2.java]: Added, here and ClusterLocalityStats.

File: xds/src/main/java/io/grpc/xds/XdsLbPolicies.java (hunk 3 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLbPolicies.java:26]: Removed EDS. Btw, our existing way of declaring all LB policy names in XdsLbPolicies is anti-pattern in Java, we may consider getting rid of it. I think using string literals in LB providers would be fine.

File: xds/src/main/java/io/grpc/xds/ClusterImplLoadBalancer.java [hunk 6]
```diff
@@ -207,7 +208,36 @@ public Subchannel createSubchannel(CreateSubchannelArgs args) {
         addresses.add(new EquivalentAddressGroup(eag.getAddresses(), attrBuilder.build()));
       }
       args = args.toBuilder().setAddresses(addresses).build();
-      return delegate().createSubchannel(args);
+      final Subchannel subchannel = delegate().createSubchannel(args);
+      Locality locality = subchannel.getAllAddresses().get(0).getAttributes().get(
+          InternalXdsAttributes.ATTR_LOCALITY);  // all addresses should be in the same locality
+      // Endpoint addresses resolved by ClusterResolverLoadBalancer should always contain
+      // attributes with its locality, including endpoints in LOGICAL_DNS clusters.
+      // In case of not (which really shouldn't), loads are aggregated under an empty locality.
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ClusterImplLoadBalancer.java:216]: Discussed offline. We would still want to keep load reporting logic in one place. From the overall design's perspective, it's more reasonable to have ClusterResolverLoadBalancer be responsible for endpoint discovery and ClusterImplLoadBalancer for applying cluster-wise configurations.

It's not a big deal for cross-policy attribute referencing.

**Comment by voidzcy** [comment]
All addressed. PTAL.

**Review by dapengzhang0 — APPROVED** [review]

File: xds/src/main/java/io/grpc/xds/ClusterImplLoadBalancer.java [hunk 7]
```diff
@@ -207,7 +208,36 @@ public Subchannel createSubchannel(CreateSubchannelArgs args) {
...
+      if (locality == null) {
+        locality = new Locality("", "", "");
+      }
+      final ClusterLocalityStats localityStats = xdsClient.addClusterLocalityStats(
+          cluster, edsServiceName, locality);
+
+      return new ForwardingSubchannel() {
+        @Override
+        public Attributes getAttributes() {
+          return delegate().getAttributes().toBuilder().set(
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/ClusterImplLoadBalancer.java:226]: Put stats object in PickSubchannelArgs instead as being pointed out offline.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ClusterImplLoadBalancer.java]: Done.

**Pull request merged**